integration/v3_lock_test.go

```diff
@@ -101,6 +101,83 @@ func TestMutexSessionRelock(t *testing.T) {
 	}
 }
 
+// TestMutexLockOnlyAfterOldestReleasesLock ensures a series of waiters
+// obtain lock only after the oldest waiter releases lock.
+// e.g a series of Waiters A <- B <- C where A holds the lock
+// and B waits before C. The failure of B won't trigger C to obtain the lock.
+func TestMutexLockOnlyAfterOldestReleasesLock(t *testing.T) {
+	clus := NewClusterV3(t, &ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	cctx, ccancel := context.WithCancel(context.Background())
+
+	cli := clus.Client(0)
+	s1, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	s2, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	s3, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+
+	m := concurrency.NewMutex(s1, "test-mutex")
+	if err := m.Lock(cctx); err != nil {
+		t.Fatal(err)
+	}
+
+	m2 := concurrency.NewMutex(s2, "test-mutex")
+	m2donec := make(chan struct{})
+	go func() {
+		defer close(m2donec)
+		m2.Lock(cctx)
+	}()
+
+	// ensure m2 waits before m3
+	time.Sleep(100 * time.Millisecond)
```

Comment:
this should watch for two PUT events on prefix "test-mutex" with a watch starting at Rev=1 (use the clientv3 watcher) instead of sleeping